src/main/java/com/orange/main/products/service/impl/ProductsServiceImpl.java

```diff
@@ -26,26 +27,44 @@ public class ProductsServiceImpl implements ProductsService {
 
     @Override
     public Products getProductById(Long productId) {
-        Optional<Products> optionalUser =  productsRepository.findById(productId);
-        return optionalUser.get();
+        Optional<Products> product =  productsRepository.findById(productId);
+        return product.get();
     }
 
     @Override
-    public Iterable<Products> getAllProducts() {
-        return productsRepository.findAll();
+    public List<Products> getAllProducts(Pageable page, List<String> fields) {
+        Page<Products> products = productsRepository.findAll(page);
+        //fields filiter
+        List<Products> rtn = products.getContent().stream().map(p -> handleUnspecifiedColumns(p, fields)).toList();
+        return rtn;
     }
 
     @Override
-    public Iterable<Products> findByNameIn(List<String> name, String sortBy, String orderBy) {
-        return productsRepository.findByNameIn(name,Sort.by("DESC".equals(orderBy)?Sort.Direction.DESC:Sort.Direction.ASC, StringUtils.isNotBlank(sortBy)?sortBy:"id"));
+    public List<Products> findByNameIn(Pageable page, List<String> name, List<String> fields) {
+        Page<Products> products = productsRepository.findByNameIn(page, name);
+        //fields filiter
+        List<Products> rtn = products.getContent().stream().map(p -> handleUnspecifiedColumns(p, fields)).toList();
+        return rtn;
     }
 
-
     public ProductsRepository getProductsRepository() {
         return this.productsRepository;
     }
 
     public void setProductsRepository(ProductsRepository productsRepository) {
         this.productsRepository = productsRepository;
     }
+
+    //fields filiter
+    public Products handleUnspecifiedColumns(Products p, List<String> specifiedColumns) {
+        if(!CollectionUtils.isEmpty(specifiedColumns)){
+            if(!specifiedColumns.contains("id")) p.setId(null);
+            if(!specifiedColumns.contains("name")) p.setName(null);
+            if(!specifiedColumns.contains("description")) p.setDescription(null);
+            if(!specifiedColumns.contains("types")) p.setTypes(null);
+            if(!specifiedColumns.contains("createdDate")) p.setCreatedDate(null);
+            if(!specifiedColumns.contains("updatedDate")) p.setUpdatedDate(null);
+        }
```

Comment:
Maybe this section needs to be refactored. It could be a disaster with more fields.

Reply:
I know, but I still think about how to do. I need some time to figure out.😓